### PR TITLE
feat: add EXISTS/NOT_EXISTS query option for specific tags filter

### PR DESCRIPTION
**Pull request by eneskaya**
This PR adds two new filters for specific tags, the `does exist/EXISTS` and `does not exist/NOT_EXISTS` filters. The `EXISTS` filter checks if a given tag key is present in a row, respectively `NOT_EXISTS` returns all rows that do not have that key present.

Screenshot 2023-04-04 at 16 36 42

### CODE REVIEW

**Comment by eneskaya** [comment]
@mlabouardy After a lot of experimenting with the queries for the NOT_EXISTS filter, I've decided to go with this solution, as it covers what we need. I've just one problem with it now, when applied the filter shows the correct results for all resources, that have tags except the one in the filter. But the `CROSS JOIN json_each(tags) ...` only shows rows, where the `tags` column is not empty.

Just to clarify, this:

```sql
SELECT
	resources.id,
	resources.resource_id,
	resources.tags
FROM
	resources
	CROSS JOIN json_each(tags)
WHERE
	TYPE = 'object'
	AND(NOT EXISTS (
			SELECT
				1 FROM json_each(resources.tags)
			WHERE (json_extract(value, '$.key') = 'Author')))
ORDER BY
	resources.id
LIMIT 50 OFFSET 0
```

shows all resources, that have tags but only if none of the keys is `'Author'`.

And this:

```sql
SELECT
	resources.id,
	resources.resource_id,
	resources.tags
FROM
	resources
WHERE
	NOT EXISTS (
			SELECT
				1 FROM json_each(resources.tags)
			WHERE (json_extract(value, '$.key') = 'Author'))
ORDER BY
	resources.id
LIMIT 50 OFFSET 0
```

shows all resources, that don't have the `Author` tag, regardless wether the `tags` column is empty, [] or have other keys in them.

The `CROSS JOIN ...` statement is added later in the code. What do you think? One would expect the second behaviour, right? If so, should I remove the `CROSS JOIN`?

File: dashboard/components/inventory/components/filter/InventoryFilterOperator.tsx [hunk 1]
```diff
@@ -22,6 +22,12 @@ const inventoryFilterOperatorOptions: InventoryFilterOperatorOptionsProps[] = [
   { label: 'is not empty', value: 'IS_NOT_EMPTY' }
 ];
 
+const inventoryFilterOperatorSpecificTagsOptions: InventoryFilterOperatorOptionsProps[] =
+  [
+    { label: 'does exist', value: 'EXISTS' },
+    { label: 'does not exist', value: 'NOT_EXISTS' }
+  ];
```
Comment (victorgaard) [inline dashboard/components/inventory/components/filter/InventoryFilterOperator.tsx:29]: The value/label also need to be on `getOperator` function of `InventoryFilterSummary` so we can display the label on the final step and on the filters tab.

In comparison:
image

image

image

File: dashboard/components/inventory/components/filter/InventoryFilter.tsx [hunk 2]
```diff
@@ -125,6 +125,8 @@ function InventoryFilter({
                         data &&
                         data.operator !== 'IS_EMPTY' &&
                         data.operator !== 'IS_NOT_EMPTY' &&
+                        data.operator !== 'EXISTS' &&
+                        data.operator !== 'NOT_EXISTS' &&
                         data.values &&
                         data.values.length === 0
                       }
```
Comment (victorgaard) [inline dashboard/components/inventory/components/filter/InventoryFilter.tsx:132]: Not sure where to drop this one, but I wanted to understand what is the expected output for when I'm filtering for a specific tag key that does not exist.

For this given case: I am filtering for specific tag key `Author` which does not exist. Is it returning all tagged items that are not `Author`? 

image

Also, it seems there are a few repeated items being returned in the response.

File: dashboard/components/inventory/components/filter/InventoryFilterOperator.tsx (hunk 1 above)
Comment (eneskaya) [reply dashboard/components/inventory/components/filter/InventoryFilterOperator.tsx:29]: ah yes good catch

File: dashboard/components/inventory/components/filter/InventoryFilter.tsx (hunk 2 above)
Comment (eneskaya) [reply dashboard/components/inventory/components/filter/InventoryFilter.tsx:132]: > For this given case: I am filtering for specific tag key Author which does not exist. Is it returning all tagged items that are not Author?

Yes, see my comment above for Mohamed.

> Also, it seems there are a few repeated items being returned in the response.

Ok, that's very odd! I'll check 🤔

Comment (eneskaya) [reply dashboard/components/inventory/components/filter/InventoryFilter.tsx:132]: Very good catch with the duplicate rows returned! I need to fix the sql query for that

Comment (victorgaard) [reply dashboard/components/inventory/components/filter/InventoryFilter.tsx:132]: Happy to help! 🙌

**Comment by mlabouardy** [comment]
> 

Yes its the expected behavior and `CROSS JOIN` could be removed in this case, could you check if this fix works when a user adds multiple filters as well

**Comment by eneskaya** [comment]
> > 
> 
> Yes its the expected behavior and `CROSS JOIN` could be removed in this case, could you check if this fix works when a user adds multiple filters as well

I added a "SELECT DISTINCT" now to filter out the duplicates. But I am really hesitant to introduce yet another `if` statement in that query code 😅 I tried removing the CROSS JOIN altogether, but then other filters stop working. So I'd have to specifically remove in the case when the user filters for NOT_EXISTS.

**Review by mlabouardy — APPROVED** [review]

**Pull request merged**